### PR TITLE
feature(chaos-mesh): use chaos-mesh for memory stress in k8s clusters

### PR DESCRIPTION
**Pull request by soyacz**
Memory stress nemesis does not work on k8s:
- #5455

This commit adds support for chaos-mesh's memory stress experiment.
Also made some bug fixing: additional experiment failure detection.
Code unification to ease adding new experiments.

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by soyacz** [comment]
staging job: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/memory-stress-eks/

File: sdcm/utils/k8s/chaos_mesh.py [hunk 1]
```diff
@@ -71,14 +71,16 @@ def initialize(self) -> None:
             self._k8s_cluster.POOL_LABEL_NAME, self._k8s_cluster.AUXILIARY_POOL_NAME)
         scylla_node_pool_affinity = get_helm_pool_affinity_values(
             self._k8s_cluster.POOL_LABEL_NAME, self._k8s_cluster.SCYLLA_POOL_NAME)
+        chaos_daemon_settings = scylla_node_pool_affinity | {
+            "runtime": "containerd", "socketPath": "/run/containerd/containerd.sock"}
```
Comment (fruch) [inline sdcm/utils/k8s/chaos_mesh.py:76]: is this correct for both EKS and kind ?

Comment (soyacz) [reply sdcm/utils/k8s/chaos_mesh.py]: actually it is not - I added code to detect runtime and adapt accordingly.

File: sdcm/nemesis.py [hunk 2]
```diff
@@ -3537,7 +3538,19 @@ def disrupt_memory_stress(self):
         Try to run stress-ng to preempt allocated memory of scylla process,
         we don't monitor swap usage in /proc/$scylla_pid/status, just make sure
         no coredump, serious db error occur during the heavy load of memory.
+
+        For k8s cluster uses chaos-mesh experiment based on https://github.com/chaos-mesh/memStress
         """
+        if self._is_it_on_kubernetes():
```
Comment (fruch) [inline sdcm/nemesis.py:3544]: Also check if it's enabled in configuration...

File: sdcm/nemesis.py [hunk 3]
```diff
@@ -3537,7 +3538,19 @@ def disrupt_memory_stress(self):
...
+            self.log.info('Try to allocate 120% available memory, the allocated memory will be swaped out')
+            experiment = MemoryStressExperiment(pod=self.target_node, duration="100s", workers=12, size="10%")
```
Comment (fruch) [inline sdcm/nemesis.py:3546]: Why are we using 12 workers here ? with stress-ng we didn't use multiple processes/threads, why do we need it here ? does it make any difference ?

Also why on each command we are using a different number of workers ?

**Review by fruch — COMMENTED** [review]
Looks good,

One question about workers 
And one configuration check missing

File: sdcm/nemesis.py (hunk 3 above)
Comment (soyacz) [reply sdcm/nemesis.py:3546]: I just followed some examples. But apparently it is not working correctly:
image

I'll try with different settings

Comment (soyacz) [reply sdcm/nemesis.py]: Using approach like with typical nodes (getting available memory from `/proc/meminfo`) is not working correctly - all the memory is not occupied:
image

My guess is that scylla container gets additional memory as needed, so getting `/proc/meminfo` value before scylla occupies all available memory leads to this error..
I'll try with percentage option for this experiment (but it accepts only up to 100%, so 120% case won't be covered) and update here.

Comment (soyacz) [reply sdcm/nemesis.py]: with 100% limit got similar result, but I found we have something like `k8s_cluster.calculated_memory_limit` which should be right - testing it.

Comment (soyacz) [reply sdcm/nemesis.py]: result with using `1.2 * k8s_cluster.calculated_memory_limit`
image
Still not all the memory is occupied - I will try few more ways.

Comment (soyacz) [reply sdcm/nemesis.py]: I tried with memory stress ramp up over 30s, similar result:
image

Comment (soyacz) [reply sdcm/nemesis.py]: Ok, I think I've found a reason for this behavior: I've read that k8s memory management kills a pod when memory usage raises to quickly. This relates to that 90% mem stress shows higher mem usage as it raises slower than for 120% stress.
When I prolonged stress ramp up to 60s, result looks like this:
image

Still not perfect, but it's getting there.

File: sdcm/nemesis.py [hunk 4]
```diff
@@ -3537,7 +3538,38 @@ def disrupt_memory_stress(self):
...
+
+        For k8s cluster uses chaos-mesh experiment based on https://github.com/chaos-mesh/memStress
         """
+        if self._is_it_on_kubernetes():
+            if not self.tester.k8s_cluster.chaos_mesh.initialized:
+                raise UnsupportedNemesis(
+                    "Chaos Mesh is not installed. Set 'k8s_use_chaos_mesh' config option to 'true'")
+            self.log.info('Try to allocate 120% available memory, the allocated memory will be swaped out')
+            memory_limit = self.tester.k8s_cluster.calculated_memory_limit
+            memory, suffix = int(memory_limit[:-2]), memory_limit[-2:]
```
Comment (vponomaryov) [inline sdcm/nemesis.py:3550]: At first, need to use existing utilities.
For this case we have `convert_memory_value_from_k8s_to_units` function.

At second, I think such conversion must be part of the `MemoryStressExperiment` class itself.
Consumer must not be bothered with such details.
So, consumer should just specified the multiplier.
Example:
```
experiment = MemoryStressExperiment(
    pod=self.target_node,
    duration="160s",
    workers=4,
    memory_multiplier=1.2,
    time_to_reach="60s",
)
```

File: sdcm/nemesis.py [hunk 5]
```diff
@@ -3537,7 +3538,38 @@ def disrupt_memory_stress(self):
...
+            for _ in range(80):
+                mem_stats = self.target_node.remoter.run("cat /sys/fs/cgroup/memory/memory.limit_in_bytes "
+                                                         "/sys/fs/cgroup/memory/memory.max_usage_in_bytes "
+                                                         "/sys/fs/cgroup/memory/memory.usage_in_bytes", verbose=False).stdout
+                LOGGER.info('mem limit/max_used/usage:\n %s', mem_stats)
+                time.sleep(2)
+            experiment.wait_until_finished()
+            self.log.info('Try to allocate 90% total memory, the allocated memory will be swaped out')
+            experiment = MemoryStressExperiment(pod=self.target_node, duration="160s",
+                                                workers=1, size=f"{memory*0.9}{suffix}", time_to_reach="60s")
```
Comment (vponomaryov) [inline sdcm/nemesis.py:3563]: Same here.

File: sdcm/utils/k8s/chaos_mesh.py [hunk 6]
```diff
@@ -99,16 +107,18 @@ class ExperimentStatus(Enum):
     UNKNOWN = 5
 
 
-class PodChaosExperiment:
-    """Base class for all PodChaos experiments."""
+class ChaosMeshExperiment:
```
Comment (vponomaryov) [inline sdcm/utils/k8s/chaos_mesh.py:110]: Is it really going to be suitable for both types of backends? K8S and VMs?

File: sdcm/utils/k8s/chaos_mesh.py [hunk 7]
```diff
@@ -166,36 +183,70 @@ def wait_until_finished(self):
...
         deep_merge(self._experiment, {
-            "kind": "PodChaos",
             "spec": {
                 "action": "pod-failure",
                 "duration": duration,
             }
         })
+
+
+class MemoryStressExperiment(ChaosMeshExperiment):
```
Comment (vponomaryov) [inline sdcm/utils/k8s/chaos_mesh.py:220]: It is K8S- specific, so, I think, it must contain something specific to K8S in it's name.

**Review by vponomaryov — CHANGES_REQUESTED** [review]

File: sdcm/utils/k8s/chaos_mesh.py (hunk 6 above)
Comment (fruch) [reply sdcm/utils/k8s/chaos_mesh.py:110]: The name change here, is that experiments would have a base class,

From where you see it should support VMs ?

File: sdcm/utils/k8s/chaos_mesh.py (hunk 7 above)
Comment (fruch) [reply sdcm/utils/k8s/chaos_mesh.py:220]: It's a good question, in theory something like memory stress we could use the chaos mesh command line, with the same parameters.

Again we might have different classes, we might implement it in the same class.

For now, we should limit it to only k8s.

We didn't yet tried the commandline in VM to understand of it's a path we want to follow.

File: sdcm/nemesis.py (hunk 4 above)
Comment (soyacz) [reply sdcm/nemesis.py:3550]: MemoryStressExperiment supports `%`: we can provide `90%` or in absolute units, so I'll stay with `%` support.
The reason I do the tricks here with memory limits read is due not supporting values >100% by chaos-mesh. 

I tried to keep the test as much as possible to VM based test. Although, I'm not sure if it is relevant, as from what I've read, K8s is rather not using swappiness (from the other hand, when reading `/sys/fs/cgroup/memory/memory.swappiness` it returns 10), is it?. And another question - will k8s allow memory to exceed the limit?
Maybe we can adapt the test and use 100% instead 120% for k8s?

File: sdcm/utils/k8s/chaos_mesh.py (hunk 7 above)
Comment (soyacz) [reply sdcm/utils/k8s/chaos_mesh.py:224]: this class inherits from `ChaosMeshExperiment`, chaos-mesh tool is designed only for K8s and this is simple basic information that must be understood by everyone using SCT. Is there special reason that we need to put `k8s` prefix for such basics?

Comment (fruch) [reply sdcm/utils/k8s/chaos_mesh.py:224]: If there an equivalent one targeted for VM, then yes we probably need some prefix.
 
But currently there isn't, so I don't think it's needed at this stage

Comment (soyacz) [reply sdcm/utils/k8s/chaos_mesh.py:224]: `Experiment` is term from chaos-mesh. When we develop function that works on all backends, we'll need to change naming completely.

File: sdcm/nemesis.py (hunk 4 above)
Comment (fruch) [reply sdcm/nemesis.py:3550]: the way we configured the pods with limit/request identical, k8s doesn't let it use more memory then available.

swap is k8s is a recently new addition (only introduce in 1.22): 
https://kubernetes.io/blog/2021/08/09/run-nodes-with-swap-alpha/

I think doing 100% for now it enough, I guess later when swap is a bit more commonly used in k8s, chaos mesh would support higher then 100%, or some other configuration values to related to swap

File: sdcm/nemesis.py (hunk 3 above)
Comment (soyacz) [reply sdcm/nemesis.py]: resutlts when getting 1gb/s ramp up:
image
Still I can see oomKill's in logs.  What I saw from VM based test results, this nemesis kills the scylla too. So, do we care? Can I leave it as it is?

Comment (fruch) [reply sdcm/nemesis.py]: oomKill that kills scylla process/pod ? that's good, i.e. the system behaves as expected, we should be aware that we'll need to add code to wait for it to recover, before moving on.

in VM we have a few issues that if we do that long enough, scylla doesn't crash but gets unresponsive, and main c-s load fails.  in some cases we only have reactor stalls.
I don't think we got OOM killed

File: sdcm/utils/k8s/chaos_mesh.py (hunk 6 above)
Comment (vponomaryov) [reply sdcm/utils/k8s/chaos_mesh.py:110]: The name is not specific to K8S, so, one may think it is suitable for any backend.

Comment (soyacz) [reply sdcm/utils/k8s/chaos_mesh.py:110]: ChaosMesh is specific to chaos-mesh which is available only on K8s.

File: sdcm/nemesis.py (hunk 3 above)
Comment (soyacz) [reply sdcm/nemesis.py]: maybe we should switch order of memory stress size: first make it 90% and then go to 100% wouldn't it be more natural?

**Comment by soyacz** [comment]
@fruch @vponomaryov it's ready for review.
I changed a bit the test flow - first 90% memory stress, then 100% - I just find it more relevant.

Last test is running, when passes I'll recover back longevity test yaml to original state.

**Comment by soyacz** [comment]
ping @fruch @vponomaryov

File: sdcm/utils/k8s/chaos_mesh.py (hunk 6 above)
Comment (fruch) [reply sdcm/utils/k8s/chaos_mesh.py:110]: The name is fine for now, 
if we ever use ChaosMesh on VMs, we can revisit the names.

File: sdcm/utils/k8s/chaos_mesh.py (hunk 7 above)
Comment (fruch) [reply sdcm/utils/k8s/chaos_mesh.py:224]: I think we can leave the name with prefixing for now, if it would be come relevant we could revisit this.

**Review by fruch — APPROVED** [review]
LGTM

**Review by vponomaryov — APPROVED** [review]

**Pull request merged**